server/src/main/java/org/elasticsearch/index/translog/TranslogHeader.java

```diff
@@ -0,0 +1,187 @@
+     * See {@link BaseTranslogReader#getFirstOperationOffset()}
+     */
+    public int sizeInBytes() {
+        return headerSizeInBytes;
+    }
+
+    static int defaultSizeInBytes(String translogUUID) {
+        return headerSize(CURRENT_VERSION, new BytesRef(translogUUID).length);
+    }
+
+    private static int headerSize(int version, int uuidLength) {
+        int size = CodecUtil.headerLength(TRANSLOG_CODEC);
+        size += Integer.BYTES + uuidLength; // uuid
+        if (version >= VERSION_PRIMARY_TERM) {
+            size += Long.BYTES;    // primary term
+            size += Integer.BYTES; // checksum
+        }
+        return size;
+    }
+
+    /**
+     * Read a translog header from the given path and file channel
+     */
+    static TranslogHeader read(final String translogUUID, final Path path, final FileChannel channel) throws IOException {
+        // This input is intentionally not closed because closing it will close the FileChannel.
+        final BufferedChecksumStreamInput in =
+            new BufferedChecksumStreamInput(new InputStreamStreamInput(java.nio.channels.Channels.newInputStream(channel), channel.size()));
+        final int version;
+        try {
+            version = CodecUtil.checkHeader(new InputStreamDataInput(in), TRANSLOG_CODEC, VERSION_CHECKSUMS, VERSION_PRIMARY_TERM);
+        } catch (CorruptIndexException | IndexFormatTooOldException | IndexFormatTooNewException e) {
+            tryReportOldVersionError(path, channel);
+            throw new TranslogCorruptedException("Translog header corrupted. path:" + path, e);
+        }
+        if (version == VERSION_CHECKSUMS) {
+            throw new IllegalStateException("pre-2.0 translog found [" + path + "]");
+        }
+        // Read the translogUUID
+        final int uuidLen = in.readInt();
+        if (uuidLen > channel.size()) {
+            throw new TranslogCorruptedException("uuid length can't be larger than the translog");
+        }
+        final BytesRef uuid = new BytesRef(uuidLen);
+        uuid.length = uuidLen;
+        in.read(uuid.bytes, uuid.offset, uuid.length);
+        final BytesRef expectedUUID = new BytesRef(translogUUID);
+        if (uuid.bytesEquals(expectedUUID) == false) {
+            throw new TranslogCorruptedException("expected shard UUID " + expectedUUID + " but got: " + uuid +
+                " this translog file belongs to a different translog. path:" + path);
+        }
+        // Read the primary term
+        final long primaryTerm;
+        if (version == VERSION_PRIMARY_TERM) {
+            primaryTerm = in.readLong();
+            assert primaryTerm >= 0 : "Primary term must be non-negative [" + primaryTerm + "]; translog path [" + path + "]";
+        } else {
+            assert version == VERSION_CHECKPOINTS : "Unknown header version [" + version + "]";
+            primaryTerm = UNKNOWN_PRIMARY_TERM;
```

Comment:
UNKOWN_PRIMARY_TERM is -1. I wonder if we should 0 as a default here. Primary terms are always non negative.

Reply:
Yes, we can use `0L` as the default. I pushed https://github.com/elastic/elasticsearch/pull/29227/commits/d50d7ed788c1d209f44ada77c2486a603c0b79aa